[Bug]: Resolve YAML raise key does not exist

From the test case of the Workflow project, it has a parallel stage to read this YAML file on a different thread.

```yaml
tmp-wf-parallel-trigger-raise-task:
  type: Workflow
  params:
    branch: str
  jobs:
    first-job:
      stages:
        - name: "Echo"
          id: raise-stage
          raise: "Raise trigger with branch: ${{ params.branch }}"

tmp-wf-parallel-trigger-raise:
  type: Workflow
  jobs:
    first-job:
      stages:
        - name: "Start run parallel stage"
          id: parallel-stage
          parallel:
            branch01:
              - name: "Stage trigger 1"
                trigger: "tmp-wf-parallel-trigger-raise-task"
                params:
                  branch: ${{ branch }}
            branch02:
              - name: "Stage trigger 2"
                trigger: "tmp-wf-parallel-trigger-raise-task"
                params:
                  branch: ${{ branch }}
```

Raise error statement:

```text
        revert = Resolver.yaml_implicit_resolvers.copy()
    
        for ch in "OoYyNn":
>           if len(Resolver.yaml_implicit_resolvers[ch]) == 1:
E           KeyError: 'O'
```

--- a/src/ddeutil/io/files.py
+++ b/src/ddeutil/io/files.py
@@ -27,6 +27,7 @@ import pickle
 import re
 from contextlib import contextmanager
 from pathlib import Path
+from threading import Lock
 from typing import (
     IO,
     Any,
@@ -38,6 +39,7 @@ from typing import (
     Protocol,
     Union,
     get_args,
+    Iterator,
 )
 
 try:
@@ -67,6 +69,7 @@ from .utils import search_env, search_env_replace
 
 logger = logging.getLogger("ddeutil.io")
 FileCompressType = Literal["gzip", "gz", "xz", "bz2"]
+LOCK: Lock = Lock()
 
 __all__: tuple[str, ...] = (
     "Fl",
@@ -237,12 +240,12 @@ class Fl(FlABC):
         )
 
     @contextmanager
-    def mopen(self, *, mode: Optional[str] = None) -> IO:
+    def mopen(self, *, mode: Optional[str] = None) -> Iterator[IO]:
         """Open with memory mode context manager.
 
-        :param mode: A opening mode that allow you to use read or write mode.
+        :param mode: An opening mode that allow you to use read or write mode.
         :type mode: Optional[str] (None)
-        :rtype: IO
+        :rtype: Iterator[IO]
         """
         mode: str = mode or "r"
         file: IO = self.open(mode=mode)
@@ -379,11 +382,11 @@ class YamlFlResolve(YamlFl):
             Handle top level yaml property ``on``
             docs: https://github.com/yaml/pyyaml/issues/696
 
-            ```
+            ```python
             import re
             from yaml.resolver import Resolver
 
-            # zap the Resolver class' internal dispatch table
+            # NOTE: zap the Resolver class' internal dispatch table
             Resolver.yaml_implicit_resolvers = {}
 
             # NOTE: Current Resolver
@@ -401,27 +404,28 @@ class YamlFlResolve(YamlFl):
                     list('tf'))
             ```
         """
-        from yaml.resolver import Resolver
+        with LOCK:
+            from yaml.resolver import Resolver
 
-        revert = Resolver.yaml_implicit_resolvers.copy()
+            revert = Resolver.yaml_implicit_resolvers.copy()
 
-        for ch in "OoYyNn":
-            if len(Resolver.yaml_implicit_resolvers[ch]) == 1:
-                del Resolver.yaml_implicit_resolvers[ch]
-            else:
-                Resolver.yaml_implicit_resolvers[ch] = [
-                    x
-                    for x in Resolver.yaml_implicit_resolvers[ch]
-                    if x[0] != "tag:yaml.org,2002:bool"
-                ]
+            for ch in "OoYyNn":
+                if len(Resolver.yaml_implicit_resolvers[ch]) == 1:
+                    del Resolver.yaml_implicit_resolvers[ch]
+                else:
+                    Resolver.yaml_implicit_resolvers[ch] = [
+                        x
+                        for x in Resolver.yaml_implicit_resolvers[ch]
+                        if x[0] != "tag:yaml.org,2002:bool"
+                    ]
 
-        with self.open(mode="r") as f:
-            rs: dict[str, Any] = yaml.load(
-                f.read(), (SafeLoader if safe else UnsafeLoader)
-            )
-            # NOTE: Override revert resolver when want to use safe load.
-            Resolver.yaml_implicit_resolvers = revert
-            return rs
+            with self.open(mode="r") as f:
+                rs: dict[str, Any] = yaml.load(
+                    f.read(), (SafeLoader if safe else UnsafeLoader)
+                )
+                # NOTE: Override revert resolver when want to use safe load.
+                Resolver.yaml_implicit_resolvers = revert
+                return rs
 
 
 class YamlEnvFl(YamlFl, EnvFlMixin):

--- a/tests/test_files_open_yaml.py
+++ b/tests/test_files_open_yaml.py
@@ -1,6 +1,8 @@
 import os
 import shutil
+import time
 from collections.abc import Generator
+from concurrent.futures.thread import ThreadPoolExecutor
 from pathlib import Path
 from textwrap import dedent
 from typing import Any
@@ -69,11 +71,30 @@ def test_read_yaml_resolve_file(target_path, yaml_str_safe, yaml_data_safe):
         f.write(yaml_str_safe)
 
     data = YamlFlResolve(path=yaml_path).read(safe=False)
-    assert "on" == data["main_key"]["sub_key"]["str2bool"]
+    assert data["main_key"]["sub_key"]["str2bool"] == "on"
     assert (
         "# Comment ${DEMO_ENV_VALUE}\n"
         "This is a long statement with comment above"
     ) == data["main_key"]["sub_key"]["statement"]
+
+
+def test_read_yaml_resolve_file_multithread(target_path, yaml_str_safe, yaml_data_safe):
+    yaml_path: Path = target_path / "test_read_file_resolve.yaml"
+    with open(yaml_path, mode="w", encoding="utf-8") as f:
+        f.write(yaml_str_safe)
+
+    def read_task():  # pragma: no cov
+        data = YamlFlResolve(path=yaml_path).read()
+        time.sleep(0.5)
+        assert data["main_key"]["sub_key"]["str2bool"] == "on"
+        assert (
+            "# Comment ${DEMO_ENV_VALUE}\n"
+            "This is a long statement with comment above"
+        ) == data["main_key"]["sub_key"]["statement"]
+
+    with ThreadPoolExecutor() as executor:
+        executor.submit(read_task)
+        read_task()
 
 
 def test_read_yaml_file_with_safe(target_path, yaml_str_safe, yaml_data_safe):
@@ -137,7 +158,7 @@ def test_read_yaml_file_with_safe_mode(
     yaml_data_env_safe,
     target_path,
 ):
-    yaml_path: str = target_path / "test_read_file_env.yaml"
+    yaml_path: Path = target_path / "test_read_file_env.yaml"
 
     with open(yaml_path, mode="w", encoding="utf-8") as f:
         yaml.dump(yaml.safe_load(yaml_str_env_safe), f)
@@ -152,7 +173,7 @@ def test_read_yaml_file_with_safe_mode_and_prepare(
     target_path,
     yaml_str_env_safe,
 ):
-    yaml_path: str = target_path / "test_read_file_env_prepare.yaml"
+    yaml_path: Path = target_path / "test_read_file_env_prepare.yaml"
 
     with open(yaml_path, mode="w", encoding="utf-8") as f:
         yaml.dump(yaml.safe_load(yaml_str_env_safe), f)
@@ -182,7 +203,7 @@ def test_read_yaml_file_with_safe_mode_and_prepare_2(
     target_path,
     yaml_str_env_safe,
 ):
-    yaml_path: str = target_path / "test_read_file_env_prepare_2.yaml"
+    yaml_path: Path = target_path / "test_read_file_env_prepare_2.yaml"
 
     with open(yaml_path, mode="w", encoding="utf-8") as f:
         yaml.dump(yaml.safe_load(yaml_str_env_safe), f)

--- a/tests/test_paths.py
+++ b/tests/test_paths.py
@@ -99,12 +99,13 @@ def make_ls(test_path: Path) -> Generator[Path, None, None]:
 def test_ls(make_ls):
     files = ls(make_ls, ignore_file=".ignore")
 
-    assert [replace_sep(str(f.relative_to(make_ls))) for f in files] == [
+    print([replace_sep(str(f.relative_to(make_ls))) for f in files])
+    assert set(replace_sep(str(f.relative_to(make_ls))) for f in files) == {
         "00_01_test.yml",
         "dir01/01_01_test.yml",
         "dir01/01_02_test.yml",
         "dir02/02_01_test.yml",
-    ]
+    }
 
 
 def test_ls_empty(test_path):

